File: src/utils/bag.ts
// Kelimeki — taş torbası
import { TILE_DATA } from '../data/tiles';
import type { Tile } from '../game/types';
import { shuffle } from './random';

/** Karıştırılmış taş torbası oluşturur. */
export function buildBag(): Tile[] {
  const b: Tile[] = [];
  for (const [letter, { pts, cnt }] of Object.entries(TILE_DATA)) {
    for (let i = 0; i < cnt; i++) b.push({ letter, pts });
  }
  return shuffle(b);
}

/**
 * Torbadan en fazla `n` taş çeker. Torbayı yerinde değiştirir
 * ve çekilen taşları döndürür.
 */
export function drawTiles(bag: Tile[], n: number): Tile[] {
  const d: Tile[] = [];
  for (let i = 0; i < n && bag.length > 0; i++) {
    d.push(bag.pop()!);
  }
  return d;
}

/** Bir harfin kalan taş dökümündeki satırı. */
export interface RemainingLetter {
  /** Ham harf ('?' joker). */
  letter: string;
  pts: number;
  /** Kalan adet. */
  count: number;
}

/** Bir taşın dağılımdaki anahtarı (joker her zaman '?'). */
function tileKey(t: Tile): string {
  return t.wild || t.letter === '?' ? '?' : t.letter;
}

/**
 * "Dışarıda" kalan taşları döndürür: tüm dağılımdan tahtadaki taşlar,
 * verilen raftaki (bakan oyuncunun) taşlar ve bu turda tahtaya konmuş ama
 * henüz onaylanmamış taşlar çıkarılır. Sonuçta torbadaki ve diğer
 * oyuncuların elindeki (görünmeyen) taşlar kalır.
 *
 * `placedTiles` ATLANMAMALI: bekleyen taşlar `PLACE_TILE` ile raftan ÇIKAR
 * ama `board`a ancak `PLAY` onayıyla yazılır (`state.placed`da beklerler) —
 * sayılmazlarsa iki kova arasındaki bu boşlukta "dışarıda" görünür ve
 * rakibin elinde sanılırlar (18 Ağustos 2026'da bildirilen hata).
 */
export function remainingTiles(
  board: (Tile | null)[][],
  myRack: Tile[],
  placedTiles: Tile[] = [],
): RemainingLetter[] {
  const counts: Record<string, number> = {};
  for (const [letter, { cnt }] of Object.entries(TILE_DATA)) counts[letter] = cnt;

  const dec = (t: Tile) => {
    const k = tileKey(t);
    if (counts[k] != null) counts[k] -= 1;
  };
  for (const row of board) for (const cell of row) if (cell) dec(cell);
  for (const t of myRack) dec(t);
  for (const t of placedTiles) dec(t);

  return Object.entries(TILE_DATA).map(([letter, { pts }]) => ({
    letter,
    pts,
    count: Math.max(0, counts[letter]),
  }));
}

File: mobile/kelimeki_core/lib/src/engine/bag.dart
// Kelimeki core — taş torbası (src/utils/bag.ts portu).
import '../data/tiles.dart';
import '../model/game_state.dart' show Board;
import '../model/tile.dart';
import '../rng.dart';

/// Karıştırılmış taş torbası. Ekleme sırası tileData sırasıdır; karıştırma
/// enjekte edilen Rng ile yapılır (TS buildBag + shuffle eşleniği).
List<Tile> buildBag(Rng rng) {
  final b = <Tile>[];
  for (final e in tileData.entries) {
    for (var i = 0; i < e.value.cnt; i++) {
      b.add(Tile(letter: e.key, pts: e.value.pts));
    }
  }
  return shuffleList(b, rng);
}

/// Torbadan en fazla n taş çeker — torbayı YERİNDE değiştirir (sondan çeker,
/// TS'teki Array.pop eşleniği).
List<Tile> drawTiles(List<Tile> bag, int n) {
  final d = <Tile>[];
  for (var i = 0; i < n && bag.isNotEmpty; i++) {
    d.add(bag.removeLast());
  }
  return d;
}

class RemainingLetter {
  final String letter;
  final int pts;
  final int count;
  const RemainingLetter({required this.letter, required this.pts, required this.count});
}

String _tileKey(Tile t) => (t.wild || t.letter == '?') ? '?' : t.letter;

/// "Dışarıda" kalan taşlar: dağılımdan tahta + bakanın rafı + bu turda
/// konmuş ama henüz onaylanmamış (`state.placed`) taşlar düşülür.
///
/// `placedTiles` ATLANMAMALI: bekleyen taşlar raftan ÇIKAR ama tahtaya ancak
/// onayla yazılır — sayılmazlarsa "dışarıda" görünüp rakibin elinde sanılırlar
/// (web `src/utils/bag.ts` ile aynı düzeltme).
List<RemainingLetter> remainingTiles(Board board, List<Tile> myRack,
    [List<Tile> placedTiles = const []]) {
  final counts = <String, int>{
    for (final e in tileData.entries) e.key: e.value.cnt
  };
  void dec(Tile t) {
    final k = _tileKey(t);
    if (counts.containsKey(k)) counts[k] = counts[k]! - 1;
  }

  for (final row in board) {
    for (final cell in row) {
      if (cell != null) dec(cell);
    }
  }
  for (final t in myRack) {
    dec(t);
  }
  for (final t in placedTiles) {
    dec(t);
  }

  return [
    for (final e in tileData.entries)
      RemainingLetter(
        letter: e.key,
        pts: e.value.pts,
        count: counts[e.key]! < 0 ? 0 : counts[e.key]!,
      ),
  ];
}
